### PR TITLE
Support tags in state strategy aliases 

### PR DESCRIPTION
**Pull request by IlyaGulya**
Now Moxy allows to specify tags in state strategy aliases.
Example:
```
@StateStrategyType(AddToEndSingleTagStrategy.class)
@interface AddToEndSingleTag {
    @StateStrategyTypeTag
    String tag() default "";
}
```

Also, there's new state strategy alias which supports it: `@AddToEndSingleTag`

### CODE REVIEW

File: moxy/src/main/java/moxy/viewstate/strategy/alias/AddToEndSingleTag.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import java.lang.annotation.Target;
+
+import moxy.viewstate.strategy.AddToEndSingleStrategy;
+import moxy.viewstate.strategy.StateStrategyType;
+import moxy.viewstate.strategy.StateStrategyTypeTag;
+
+/**
+ * State strategy alias for {@link moxy.viewstate.strategy.AddToEndSingleTagStrategy}.
+ * Applying this annotation has the same effect as applying
+ * {@code @StateStrategyType(AddToEndSingleTagStrategy.class, tag = "someTag)}.
```
Comment (alaershov) [inline moxy/src/main/java/moxy/viewstate/strategy/alias/AddToEndSingleTag.java]: looks like the quite is missing after "someTag"

File: moxy-compiler/src/main/java/moxy/compiler/viewstate/ViewInterfaceProcessor.kt [hunk 2]
```diff
@@ -261,30 +247,110 @@ class ViewInterfaceProcessor(
...
+        val enclosed =
+            listOfNotNull(element.getAnnotationMirror(StateStrategyType::class))
+                .map { it.toStrategyWithTag(executableElement) }
+
+        val methodAnnotations =
+            element
+                .annotationMirrors
+
+        val aliased =
+            methodAnnotations
```
Comment (alaershov) [inline moxy-compiler/src/main/java/moxy/compiler/viewstate/ViewInterfaceProcessor.kt]: Looks scary) Maybe refactor this into smaller functions? Or add some comments at least? Or both :)

File: moxy-compiler/src/main/java/moxy/compiler/viewstate/ViewInterfaceProcessor.kt [hunk 3]
```diff
@@ -261,30 +247,110 @@ class ViewInterfaceProcessor(
...
+        elementUtils.getTypeElement(String::class.java.name)
+    }
+
+    private fun getStateStrategy(element: Element): StrategyWithTag? {
+        val executableElement = element as? ExecutableElement
+        val enclosed =
+            listOfNotNull(element.getAnnotationMirror(StateStrategyType::class))
+                .map { it.toStrategyWithTag(executableElement) }
+
+        val methodAnnotations =
```
Comment (alaershov) [inline moxy-compiler/src/main/java/moxy/compiler/viewstate/ViewInterfaceProcessor.kt]: Is this code style common in Moxy? Is it worth it to add a new line after each word?

File: moxy-compiler/src/test/java/moxy/compiler/StrategyAliasTest.kt [hunk 4]
```diff
@@ -92,6 +92,141 @@ class StrategyAliasTest : CompilerTest() {
...
+                String tag2() default "";
+            }
+            
+            public interface ViewInterface extends MvpView {
+                @OneExecution
+                void testMethod();
+            }
+        """.toJavaFile()
+
+        val presenter = generateViewStateFor("moxy.ViewInterface")
```
Comment (alaershov) [inline moxy-compiler/src/test/java/moxy/compiler/StrategyAliasTest.kt]: Why viewState is called presenter here?

File: moxy-compiler/src/test/java/moxy/compiler/StrategyAliasTest.kt [hunk 5]
```diff
@@ -92,6 +92,141 @@ class StrategyAliasTest : CompilerTest() {
...
+        val compilation = compileSourcesWithProcessor(viewInterface, presenter)
+
+        compilation.assertThatIt()
+            .hadErrorContaining("""
+                Annotation @StateStrategyTypeTag can't be applied to more than one method!
```
Comment (alaershov) [inline moxy-compiler/src/test/java/moxy/compiler/StrategyAliasTest.kt]: raw string is probably not neccessary in this case, it's just a single line

**Review by alaershov — CHANGES_REQUESTED** [review]
Looks good, but I'd like you to fix some code style issues before merging to dev

File: moxy/src/main/java/moxy/viewstate/strategy/alias/AddToEndSingleTag.java (hunk 1 above)
Comment (IlyaGulya) [reply moxy/src/main/java/moxy/viewstate/strategy/alias/AddToEndSingleTag.java]: Yep, fixed

File: moxy-compiler/src/main/java/moxy/compiler/viewstate/ViewInterfaceProcessor.kt (hunk 2 above)
Comment (IlyaGulya) [reply moxy-compiler/src/main/java/moxy/compiler/viewstate/ViewInterfaceProcessor.kt]: There's not that much to extract.
I've extracted a separate function which searching aliased state strategies.
Also extracted an extension method which is searching element methods.

File: moxy-compiler/src/main/java/moxy/compiler/viewstate/ViewInterfaceProcessor.kt (hunk 3 above)
Comment (IlyaGulya) [reply moxy-compiler/src/main/java/moxy/compiler/viewstate/ViewInterfaceProcessor.kt]: Sorry, that is code style I'm using on all my projects. Fixed to match Moxy's one.

File: moxy-compiler/src/test/java/moxy/compiler/StrategyAliasTest.kt (hunk 5 above)
Comment (IlyaGulya) [reply moxy-compiler/src/test/java/moxy/compiler/StrategyAliasTest.kt]: fixed

File: moxy-compiler/src/test/java/moxy/compiler/StrategyAliasTest.kt (hunk 4 above)
Comment (IlyaGulya) [reply moxy-compiler/src/test/java/moxy/compiler/StrategyAliasTest.kt]: That was a copy-paste.
This issue exists in all tests in this file.
Fixed all of them.

**Review by aasitnikov — CHANGES_REQUESTED** [review]
Hi, Ilya, thanks for PR! 

I like the proposed feature - it's good and useful. There was some nice improvements in codebase, new tests cover a lot of cases. But it seems to me that feature is not properly designed in first place. Let me explain.

Current version has "strategy aliases" feature - if some annotation is annotated with `@StateStrategyType`, then element in view interface (be it a method or interface itself), annotated with this alias annotation should behave the same as if it was annotated with `@StateStrategyType` itself. Moxy library bundles some predefined alias annotations, but this feature does not constrain these annotations to be only in the library itself - users can define custom alias annotations for custom `StateStrategy` classes. Also, current Moxy annotation processor implementation does not take alias annotation's methods into account, so these methods have no effect on the way processor chooses strategy. This makes it hard to use strategy aliases on strategies with tags, as one will have to create new alias annotation for every new tag. So we want to make it easy to specify tags in alias annotations on call site, like this:

```
interface SomeViewInterface : MvpView {

    @AddToEndSingleTag(tag = "state")
    fun showContent()
}
```

As far as I understand, your PR adds new `@StateStrategyTypeTag`, which must be on at most one method in alias annotation. Annotated method will be a provider for generated method command tag. I think it is unnecessary and kind of boilerplate.

So I propose different solution - forbid alias annotations from containing any methods, with exception of `String tag()`. If this method is present, then tag specified in this annotation will be used as a tag for generated method command. 

This change raises two problems. First - if someone created alias annotation with only one method `String tag()`, then program behaviour would change, without compilation error. Second - what the tag value will be, if both alias annotation and `@StateStrategyType` annotation contain tags? 

As for me, I think first scenario is highly unlikely, and we can write about it in changelog, noting that new release contains breaking changes regarding custom alias annotations. And for second, I propose to forbid a tag on `@StateStrategyType`, if alias contains `tag()` method:

```
// valid - tag inferred from @AddToEndSingleTag
@StateStrategyType(AddToEndSingleStrategy.class)
public @interface AddToEndSingleTag {

    String tag() default "";
}

// valid - tag is inferred form @StateStrategyType
@StateStrategyType(value = AddToEndSingleStrategy.class, tag = "customTag")
public @interface AddToEndSingleWithCustomTag {
}

// compilation error - you can't specify tags twice
@StateStrategyType(value = AddToEndSingleStrategy.class, tag = "customTag")
public @interface AddToEndSingleTag {

    String tag() default "";
}
```

**Pull request closed without merging**